fuel-vm/src/interpreter/executors/main.rs

```diff
@@ -42,52 +44,113 @@ impl<T> Interpreter<PredicateStorage, T> {
     /// This is not a valid entrypoint for debug calls. It will only return a `bool`, and not the
     /// VM state required to trace the execution steps.
     pub fn check_predicates<Tx>(
-        checked: Checked<Tx>,
+        checked: &Checked<Tx>,
         params: ConsensusParameters,
         gas_costs: GasCosts,
     ) -> Result<PredicatesChecked, PredicateVerificationFailed>
     where
         Tx: ExecutableTransaction,
-        <Tx as IntoChecked>::Metadata: CheckedMetadata,
+        <Tx as IntoChecked>::CheckedMetadata: CheckedMetadata,
     {
         if !checked.transaction().check_predicate_owners(&params) {
             return Err(PredicateVerificationFailed::InvalidOwner);
         }
 
         let mut vm = Interpreter::with_storage(PredicateStorage::default(), params, gas_costs);
 
+        let mut cumulative_gas_used: Word = 0;
+
+        vm.init_predicate(checked.clone());
+
         // Needed for now because checked is only freed once the value is collected into a Vec
         #[allow(clippy::needless_collect)]
         let predicates: Vec<_> = (0..checked.transaction().inputs().len())
             .filter_map(|i| RuntimePredicate::from_tx(&params, checked.transaction(), i))
             .collect();
 
-        // Since we reuse the vm objects otherwise, we need to keep the actual gas here
-        let tx_gas_limit = checked.transaction().limit();
-        let mut remaining_gas = tx_gas_limit;
-
-        vm.init_predicate(checked);
-
         for predicate in predicates {
             // VM is cloned because the state should be reset for every predicate verification
             let mut vm = vm.clone();
 
-            vm.context = Context::Predicate { program: predicate };
-            vm.set_remaining_gas(remaining_gas);
+            let gas_used = predicate.gas_used();
+            vm.context = Context::PredicateVerification { program: predicate };
+            vm.set_gas(gas_used);
 
             if !matches!(vm.verify_predicate()?, ProgramState::Return(0x01)) {
                 return Err(PredicateVerificationFailed::False);
             }
 
-            remaining_gas = vm.registers[RegId::GGAS];
+            if vm.registers[RegId::GGAS] != 0 {
+                return Err(PredicateVerificationFailed::GasMismatch);
+            }
+            cumulative_gas_used = cumulative_gas_used
+                .checked_add(gas_used)
+                .expect("cumulative gas overflow");
```

Comment:
We don't want expects in production code. This should return an error instead.

Reply:
removed and returning an error now